src/hotspot/share/opto/vectorIntrinsics.cpp

```diff
@@ -760,12 +760,29 @@ bool LibraryCallKit::inline_vector_mem_operation(bool is_store) {
   SafePointNode* old_map = clone_map();
 
   Node* addr = make_unsafe_address(base, offset, (is_mask ? T_BOOLEAN : elem_bt), true);
+
+  // The memory barrier checks are based on ones for unsafe access.
+  // This is not 1-1 implementation.
+  const Type *const base_type = gvn().type(base);
+
+  // Is off heap access (true implies can_access_non_heap = true)
+  const bool off_heap_access = TypePtr::NULL_PTR == base_type;
+
   // Can base be NULL? Otherwise, always on-heap access.
-  bool can_access_non_heap = TypePtr::NULL_PTR->higher_equal(gvn().type(base));
+  const bool can_access_non_heap = TypePtr::NULL_PTR->higher_equal(base_type);
+
+  // Not determined access, base can and can not be null.
+  const bool mixed_access = !off_heap_access && can_access_non_heap;
+  bool mismatched_array = false;
 
   const TypePtr *addr_type = gvn().type(addr)->isa_ptr();
   const TypeAryPtr* arr_type = addr_type->isa_aryptr();
 
+  if (arr_type == NULL && !off_heap_access) {
```

Comment:
Not sure this case can occur, since we only pass array instances or `null` and we cannot pollute with different types of array. Might need some guidance from a HotSpot expert? Perhaps we can remove this code or place in an assert?

We do have a form of mismatched access between `char[]` and `ShortVector` and `boolean[]` and `ByteVector` but that's a different case.